File: desktop/src/renderer/domain/shortCoach.ts
import { z } from "zod";
import type { ContactDraft, Opportunity } from "./models";
import { explicitInstant } from "./opportunityLibrary";

const id = z.string().trim().min(1).max(128);
const body = z
  .string()
  .min(1)
  .max(8000)
  .refine((v) => !!v.trim() && !v.includes("\0"));
const hash = z.string().regex(/^[a-f0-9]{64}$/);
const instant = z.string().refine(explicitInstant);
const accountScopeSchema = z
  .object({ id, version: z.number().int().min(1) })
  .strict();
export type CoachAccountScope = z.infer<typeof accountScopeSchema>;
const url = z
  .string()
  .max(2048)
  .refine((value) => {
    try {
      const parsed = new URL(value);
      return (
        ["https:", "http:"].includes(parsed.protocol) &&
        !parsed.username &&
        !parsed.password
      );
    } catch {
      return false;
    }
  });
export const COACH_PURPOSES = {
  requirement: "确认具体需求",
  materials: "询问资料获取方式",
  scope: "确认服务范围",
} as const;
export type CoachPurpose = keyof typeof COACH_PURPOSES;
export const coachBindingSchema = z
  .object({
    accountScope: accountScopeSchema,
    requestId: id,
    opportunityId: id,
    profileVersionId: id,
    sourceEvidenceVersion: id,
    sourceUrl: url,
    sourceObservedAt: instant,
    channel: z.enum(["comment", "dm"]),
    draftVersion: z.number().int().min(1),
    draftHash: hash,
    purpose: z.enum(["requirement", "materials", "scope"]),
  })
  .strict();
export type CoachBinding = z.infer<typeof coachBindingSchema>;
export interface CoachInput {
  binding: CoachBinding;
  content: string;
  sourceText: string;
}
const quoteSchema = z
  .object({
    id,
    text: body,
    start: z.number().int().min(0),
    end: z.number().int().min(1),
    sourceUrl: url,
    sourceEvidenceVersion: id,
  })
  .strict();
export const coachSuggestionSchema = z
  .object({
    suggestionId: id,
    binding: coachBindingSchema,
    content: body,
    question: body,
    context: z
      .object({ summary: body, quoteIds: z.array(id).min(1).max(12) })
      .strict(),
    quotes: z.array(quoteSchema).min(1).max(12),
    checks: z
      .array(
        z
          .object({
            kind: z.enum(["CONTEXT", "ONE_QUESTION", "PROMISE", "LENGTH"]),
            status: z.enum(["SUPPORTED", "NEEDS_REVIEW"]),
            message: body,
            quoteIds: z.array(id).max(12),
          })
          .strict(),
      )
      .max(12),
    createdAt: instant,
    expiresAt: instant,
  })
  .strict();
export type CoachSuggestion = z.infer<typeof coachSuggestionSchema>;
export async function textDigest(value: string) {
  const digest = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(value),
  );
  return Array.from(new Uint8Array(digest), (n) =>
    n.toString(16).padStart(2, "0"),
  ).join("");
}
export function coachSourceKey(row: Opportunity) {
  return JSON.stringify([
    row.id,
    row.profileVersionId,
    row.profileStatus,
    row.sourceStatus,
    row.sourceEvidenceVersion,
    row.sourceObservedAt,
    row.url,
    row.excerpt,
    row.sample === true,
  ]);
}
export function coachSourceProblem(row: Opportunity) {
  if (row.sample || row.id === "sample")
    return "公开研究样例只读，不能调用客户短句服务。";
  if (row.profileStatus !== "CONFIRMED" || !row.profileVersionId)
    return "画像版本尚未确认，请先核对商机画像。";
  if (
    row.sourceStatus !== "OPEN" ||
    !row.sourceEvidenceVersion ||
    !row.sourceObservedAt ||
    !row.excerpt.trim()
  )
    return "来源或证据版本尚未核验，请刷新原文后再生成短句建议。";
  if (
    !id.safeParse(row.profileVersionId).success ||
    !id.safeParse(row.sourceEvidenceVersion).success ||
    !instant.safeParse(row.sourceObservedAt).success ||
    !url.safeParse(row.url).success ||
    !body.safeParse(row.excerpt).success
  )
    return "来源证据格式无效，请重新核验。";
  return "";
}
export async function makeCoachInput(
  row: Opportunity,
  draft: ContactDraft,
  purpose: CoachPurpose,
  requestId: string,
  accountScope: CoachAccountScope,
): Promise<CoachInput> {
  const problem = coachSourceProblem(row);
  if (problem) throw new Error(problem);
  if (
    draft.opportunityId !== row.id ||
    draft.content.length > 8000 ||
    draft.content.includes("\0")
  )
    throw new Error("草稿与商机不一致或内容超出限制。");
  const binding = coachBindingSchema.parse({
    accountScope,
    requestId,
    opportunityId: row.id,
    profileVersionId: row.profileVersionId,
    sourceEvidenceVersion: row.sourceEvidenceVersion,
    sourceUrl: row.url,
    sourceObservedAt: row.sourceObservedAt,
    channel: draft.channel,
    draftVersion: draft.version,
    draftHash: await textDigest(draft.content),
    purpose,
  });
  return { binding, content: draft.content, sourceText: row.excerpt };
}
export function readCoachSuggestion(
  value: unknown,
  input: CoachInput,
  now = Date.now(),
) {
  const data = coachSuggestionSchema.parse(value);
  if (
    JSON.stringify(data.binding) !==
    JSON.stringify(coachBindingSchema.parse(input.binding))
  )
    throw new Error("短句建议与本次草稿、来源或请求不匹配，当前内容已保留。");
  if (
    Date.parse(data.expiresAt) <= now ||
    Date.parse(data.createdAt) > now + 60_000 ||
    Date.parse(data.expiresAt) <= Date.parse(data.createdAt)
  )
    throw new Error("短句建议已过期或时间无效，请重新生成。");
  const quoteIds = new Set(data.quotes.map((quote) => quote.id));
  if (
    quoteIds.size !== data.quotes.length ||
    data.quotes.some(
      (quote) =>
        quote.sourceUrl !== input.binding.sourceUrl ||
        quote.sourceEvidenceVersion !== input.binding.sourceEvidenceVersion ||
        quote.end > input.sourceText.length ||
        input.sourceText.slice(quote.start, quote.end) !== quote.text,
    )
  )
    throw new Error("建议引用与当前原文版本不一致，不能应用。");
  if (
    [
      ...data.context.quoteIds,
      ...data.checks.flatMap((check) => check.quoteIds),
    ].some((quoteId) => !quoteIds.has(quoteId))
  )
    throw new Error("建议依据缺少对应原文引用，不能应用。");
  if (
    data.checks.some(
      (check) =>
        ["CONTEXT", "PROMISE"].includes(check.kind) &&
        check.status === "SUPPORTED" &&
        !check.quoteIds.length,
    )
  )
    throw new Error("上下文或承诺检查缺少原文依据，不能标记为有依据。");
  if (!data.content.includes(data.question))
    throw new Error("建议问题与完整短句不一致，不能应用。");
  return data;
}

export const draftSaveBindingSchema = z
  .object({
    opportunityId: id,
    channel: z.enum(["comment", "dm"]),
    requestId: id,
    contentHash: hash,
  })
  .strict();
export type DraftSaveBinding = z.infer<typeof draftSaveBindingSchema>;
const contactDraftSchema = z
  .object({
    opportunityId: id,
    channel: z.enum(["comment", "dm"]),
    content: body,
    savedContent: z.string().max(8000),
    version: z.number().int().min(1),
    accountId: z.string().max(512),
    recipient: z.string().max(512),
    confirmedFingerprint: z.string().optional(),
  })
  .strict();
export const draftSnapshotSchema = z
  .object({
    draft: contactDraftSchema,
    accountScope: accountScopeSchema.nullable(),
    profileVersionId: z.string().max(128),
    sourceEvidenceVersion: z.string().max(128).nullable(),
  })
  .strict();
export type DraftSnapshot = z.infer<typeof draftSnapshotSchema>;
export interface DraftSaveInput {
  binding: DraftSaveBinding;
  snapshot: DraftSnapshot;
}
export const draftSaveReceiptSchema = z
  .object({
    binding: draftSaveBindingSchema,
    status: z.enum(["SUCCEEDED", "FAILED", "PENDING", "UNKNOWN"]),
    confirmed: z.boolean(),
    snapshot: draftSnapshotSchema.optional(),
  })
  .strict();
export type DraftSaveReceipt = z.infer<typeof draftSaveReceiptSchema>;
export function draftSaveKey(binding: DraftSaveBinding) {
  const b = draftSaveBindingSchema.parse(binding);
  return JSON.stringify([
    b.opportunityId,
    b.channel,
    b.requestId,
    b.contentHash,
  ]);
}
export function draftSaveBindingFromKey(key: string) {
  const parsed: unknown = JSON.parse(key);
  if (!Array.isArray(parsed) || parsed.length !== 4)
    throw new Error("草稿保存记录无效，请先核对原操作。");
  return draftSaveBindingSchema.parse({
    opportunityId: parsed[0],
    channel: parsed[1],
    requestId: parsed[2],
    contentHash: parsed[3],
  });
}
export function draftSnapshot(
  row: Opportunity,
  draft: ContactDraft,
  accountScope?: CoachAccountScope,
): DraftSnapshot {
  return draftSnapshotSchema.parse({
    draft: { ...draft, confirmedFingerprint: undefined },
    accountScope: accountScope || null,
    profileVersionId: row.profileVersionId,
    sourceEvidenceVersion: row.sourceEvidenceVersion || null,
  });
}
export function snapshotDigest(snapshot: DraftSnapshot) {
  const { draft, accountScope, profileVersionId, sourceEvidenceVersion } =
    draftSnapshotSchema.parse(snapshot);
  return textDigest(
    JSON.stringify([
      draft.opportunityId,
      draft.channel,
      draft.version,
      draft.content,
      draft.accountId,
      draft.recipient,
      profileVersionId,
      sourceEvidenceVersion,
      accountScope,
    ]),
  );
}
export async function readDraftSaveReceipt(
  value: unknown,
  binding: DraftSaveBinding,
) {
  const receipt = draftSaveReceiptSchema.parse(value);
  if (draftSaveKey(receipt.binding) !== draftSaveKey(binding))
    throw new Error("保存回执与原请求不匹配，保护继续保留。");
  if (["SUCCEEDED", "FAILED"].includes(receipt.status) && !receipt.confirmed)
    throw new Error("保存终态尚未确认，保护继续保留。");
  if (
    receipt.status === "SUCCEEDED" &&
    (!receipt.snapshot ||
      receipt.snapshot.draft.savedContent !== receipt.snapshot.draft.content ||
      (await snapshotDigest(receipt.snapshot)) !== binding.contentHash)
  )
    throw new Error("已保存内容与原提交快照不一致，保护继续保留。");
  return receipt;
}

File: pilot/short_coach.py
"""Owner-isolated short-coach admission, grounding and immutable replay."""
from __future__ import annotations

from datetime import datetime, timezone, timedelta
import hashlib, json, re, threading
from typing import Annotated, Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field, AfterValidator, model_validator

from pilot.auth import TokenClaims
from pilot.outreach_contract import canonical_uuid
from pilot.sessions import PilotSessionRegistry

POLICY_VERSION="short-coach-public-draft-v1"

class ShortCoachError(Exception):
    _status={"invalid_request":422,"invalid_session":401,"capability_unavailable":501,
        "disclosure_mismatch":409,"short_coach_request_conflict":409,"short_coach_processing":409,
        "short_coach_failed":502,"short_coach_quota_exceeded":429,"short_coach_result_expired":409}
    def __init__(self,code,status=None): self.code,self.status=code,status or self._status.get(code,409); super().__init__(code)

def _text(value):
    value.encode("utf-8")
    if "\0" in value: raise ValueError
    return value
Id=Annotated[str,AfterValidator(canonical_uuid)]
Text=Annotated[str,Field(max_length=8000),AfterValidator(_text)]

class _Strict(BaseModel): model_config=ConfigDict(strict=True,extra="forbid",hide_input_in_errors=True)
class AccountScope(_Strict): id:Id; version:Annotated[int,Field(ge=1,le=1)]
class CoachBinding(_Strict):
    accountScope:AccountScope; requestId:Id; opportunityId:Id; profileVersionId:Id
    sourceEvidenceVersion:Id; sourceUrl:str=Field(min_length=1,max_length=2048)
    sourceObservedAt:str=Field(min_length=1,max_length=64); channel:Literal["comment","dm"]
    draftVersion:Annotated[int,Field(ge=1,le=2147483647)]; draftHash:str=Field(pattern=r"^[a-f0-9]{64}$")
    purpose:Literal["requirement","materials","scope"]
class CoachInput(_Strict):
    binding:CoachBinding; content:Text; sourceText:Text=Field(min_length=1)
    @model_validator(mode="after")
    def bound(self):
        if hashlib.sha256(self.content.encode()).hexdigest()!=self.binding.draftHash: raise ValueError("draft hash")
        try:
            u=self.binding.sourceUrl
            if not (u.startswith("https://") or u.startswith("http://")) or "@" in u.split("/",3)[2]: raise ValueError
            d=datetime.fromisoformat(self.binding.sourceObservedAt.replace("Z","+00:00"))
            if d.tzinfo is None: raise ValueError
        except Exception: raise ValueError("source binding") from None
        return self
class Disclosure(_Strict):
    accepted:Literal[True]; inputHash:str=Field(pattern=r"^[a-f0-9]{64}$")
    modelProvider:str; modelName:str; policyVersion:Literal[POLICY_VERSION]
class GenerateInput(CoachInput): disclosure:Disclosure

def _hash(value):
    b=value.binding
    fields=[b.accountScope.id,b.accountScope.version,b.requestId,b.opportunityId,b.profileVersionId,
        b.sourceEvidenceVersion,b.sourceUrl,b.sourceObservedAt,b.channel,b.draftVersion,b.draftHash,b.purpose,
        value.content,value.sourceText]
    return hashlib.sha256(json.dumps(fields,ensure_ascii=False,separators=(",",":"),allow_nan=False).encode()).hexdigest()

def _utf16_offset(text,index): return len(text[:index].encode("utf-16-le"))//2

def _same_millisecond(left, right):
    def millis(value):
        parsed=datetime.fromisoformat(value.replace("Z","+00:00"))
        return int(parsed.timestamp()*1000)
    try: return millis(left)==millis(right)
    except (ValueError,TypeError,OverflowError): return False

def build_suggestion(raw,result):
    value=CoachInput.model_validate(raw); source=value.sourceText
    if type(result) is not dict or set(result)!={"content","question","quote"}: raise ValueError
    content,question,quote=(result[k] for k in ("content","question","quote"))
    if not all(type(v) is str and v and "\0" not in v for v in (content,question,quote)): raise ValueError
    for item in (content,question,quote): item.encode("utf-8")
    if (len(content)>120 or content.count("?")+content.count("？")!=1 or content.count(question)!=1
            or not question.endswith(("?","？")) or question.count("?")+question.count("？")!=1): raise ValueError
    start=source.find(quote)
    if start<0: raise ValueError
    qid=str(uuid4()); now=datetime.now(timezone.utc); b=value.binding
    quotes=[{"id":qid,"text":quote,"start":_utf16_offset(source,start),"end":_utf16_offset(source,start+len(quote)),
        "sourceUrl":b.sourceUrl,"sourceEvidenceVersion":b.sourceEvidenceVersion}]
    return {"suggestionId":str(uuid4()),"binding":b.model_dump(),"content":content,"question":question,
        "context":{"summary":"建议仅依据当前公开原文与人工草稿。","quoteIds":[qid]},"quotes":quotes,
        "checks":[{"kind":"CONTEXT","status":"SUPPORTED","message":"上下文有原文引用。","quoteIds":[qid]},
          {"kind":"ONE_QUESTION","status":"SUPPORTED","message":"短句仅含一个问题。","quoteIds":[]},
          {"kind":"PROMISE","status":"NEEDS_REVIEW","message":"服务承诺须人工复核。","quoteIds":[]},
          {"kind":"LENGTH","status":"SUPPORTED","message":"短句不超过120个字符。","quoteIds":[]}],
        "createdAt":now.isoformat().replace("+00:00","Z"),"expiresAt":(now+timedelta(minutes=5)).isoformat().replace("+00:00","Z")}

class ShortCoachService:
    def __init__(self,database,model=None):
        self.database,self.model=database,model
        self.sessions=PilotSessionRegistry(database) if database else None
        self._call_lock=threading.Lock()
    def _configured(self):
        if self.model is None or not getattr(self.model,"available",False): raise ShortCoachError("capability_unavailable")
        provider,name=getattr(self.model,"provider",None),getattr(self.model,"model",None)
        if (type(provider) is not str or not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.-]{0,79}",provider)
                or type(name) is not str or not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.:/-]{0,199}",name)):
            raise ShortCoachError("capability_unavailable")
        return provider,name
    def preview_unverified(self,raw):
        value=CoachInput.model_validate(raw); provider,name=self._configured()
        return {"inputHash":_hash(value),"modelProvider":provider,"modelName":name,"policyVersion":POLICY_VERSION}
    def _facts(self,cursor,claims,value):
        tenant=self.sessions.require_active(cursor,claims)
        if value.binding.accountScope.id!=tenant: raise ShortCoachError("invalid_session",403)
        cursor.execute("SELECT o.profile_version_id,o.source_status,o.intent_status,p.status,s.health,e.payload,e.payload_sha256 "
            "FROM pilot_opportunities o JOIN business_profile_versions p ON p.tenant_id=o.tenant_id AND p.profile_version_id=o.profile_version_id "
            "JOIN pilot_sources s ON s.tenant_id=o.tenant_id AND s.source_id=o.source_id LEFT JOIN pilot_opportunity_evidence e "
            "ON e.tenant_id=o.tenant_id AND e.opportunity_id=o.opportunity_id WHERE o.tenant_id=%s AND o.opportunity_id=%s FOR SHARE OF o,p,s",
            (tenant,value.binding.opportunityId))
        row=cursor.fetchone()
        if row is None: raise ShortCoachError("opportunity_not_found",404)
        if (row[0]!=value.binding.profileVersionId or row[1]!="OPEN" or row[2]=="CLOSED"
                or row[3]!="CONFIRMED" or row[4]!="OPEN"): raise ShortCoachError("short_coach_facts_changed")
        from pilot.opportunity_evidence import evidence_view
        proof=evidence_view(row[5],row[6],opportunity_id=value.binding.opportunityId,profile_version_id=value.binding.profileVersionId)
        try:
            snap=proof["snapshot"]; src=snap["source"]; obs=snap["observation"]
            same_time=_same_millisecond(obs["observed_at"],value.binding.sourceObservedAt)
            if proof["status"]!="CAPTURED" or src["body"]!=value.sourceText or src["version_id"]!=value.binding.sourceEvidenceVersion or src["public_url"]!=value.binding.sourceUrl or not same_time: raise ValueError
        except Exception: raise ShortCoachError("short_coach_facts_changed") from None
        return tenant
    def preview(self,claims,raw):
        try: value=CoachInput.model_validate(raw)
        except Exception: raise ShortCoachError("invalid_request") from None
        with self.database.connect() as conn,conn.cursor() as cursor: self._facts(cursor,claims,value)
        return self.preview_unverified(value.model_dump())
    def generate(self,claims,raw):
        try: value=GenerateInput.model_validate(raw)
        except Exception: raise ShortCoachError("invalid_request") from None
        provider,name=self._configured(); expected=self.preview_unverified(value.model_dump(exclude={"disclosure"}))
        if value.disclosure.model_dump()!={"accepted":True,**expected}: raise ShortCoachError("disclosure_mismatch")
        request_hash=expected["inputHash"]; b=value.binding
        with self.database.connect() as conn,conn.cursor() as cursor:
            tenant=self._facts(cursor,claims,value)
            lock=int.from_bytes(hashlib.sha256(f"short-coach-v1\0{tenant}".encode()).digest()[:8],"big",signed=True)
            cursor.execute("SELECT pg_advisory_xact_lock(%s)",(lock,))
            cursor.execute("SELECT request_hash,state,result,error_code,model_provider,model_name,created_at FROM pilot_short_coach_requests WHERE tenant_id=%s AND owner_user_id=%s AND request_id=%s",(tenant,claims.user_id,b.requestId))
            old=cursor.fetchone()
            if old:
                if old[0]!=request_hash or old[4:6]!=(provider,name): raise ShortCoachError("short_coach_request_conflict")
                if old[1]=="PROCESSING": raise ShortCoachError("short_coach_processing")
                if old[1]=="FAILED": raise ShortCoachError(old[3] or "short_coach_failed",502)
                if old[6] < datetime.now(timezone.utc)-timedelta(minutes=5): raise ShortCoachError("short_coach_result_expired")
                return old[2]
            if not self._call_lock.acquire(blocking=False): raise ShortCoachError("short_coach_processing")
            try:
                cursor.execute("INSERT INTO pilot_short_coach_daily_quota(tenant_id,quota_day,call_count) VALUES(%s,(clock_timestamp() AT TIME ZONE 'UTC')::date,1) "
                    "ON CONFLICT(tenant_id,quota_day) DO UPDATE SET call_count=pilot_short_coach_daily_quota.call_count+1 "
                    "WHERE pilot_short_coach_daily_quota.call_count<50 RETURNING call_count",(tenant,))
                if cursor.fetchone() is None: raise ShortCoachError("short_coach_quota_exceeded")
                cursor.execute("INSERT INTO pilot_short_coach_requests(tenant_id,owner_user_id,request_id,opportunity_id,request_hash,model_provider,model_name,state) VALUES(%s,%s,%s,%s,%s,%s,%s,'PROCESSING')",
                    (tenant,claims.user_id,b.requestId,b.opportunityId,request_hash,provider,name))
            except Exception:
                self._call_lock.release()
                raise
        try:
            try:
                result=build_suggestion(value.model_dump(exclude={"disclosure"}),self.model.generate(sourceText=value.sourceText,content=value.content,channel=b.channel,purpose=b.purpose))
                state,error="SUCCEEDED",None
            except Exception:
                result,state,error=None,"FAILED","short_coach_failed"
            with self.database.connect() as conn,conn.cursor() as cursor:
                self.sessions.require_active(cursor,claims); self._facts(cursor,claims,value)
                cursor.execute("UPDATE pilot_short_coach_requests SET state=%s,result=%s::jsonb,error_code=%s,updated_at=clock_timestamp() WHERE tenant_id=%s AND owner_user_id=%s AND request_id=%s AND state='PROCESSING'",
                    (state,None if result is None else json.dumps(result,ensure_ascii=False),error,tenant,claims.user_id,b.requestId))
            if result is None: raise ShortCoachError(error,502)
            return result
        finally:
            self._call_lock.release()
